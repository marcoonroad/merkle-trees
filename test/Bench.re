open Core_bench.Bench;
module Command = Core.Command;
module PRNG = Support.PRNG;
module Merkle = Support.MerkleTree;

let heights = [3, 5, 10, 12, 15];

let power_of_two = number => {
  int_of_float(2.0 ** float_of_int(number));
};

let generate = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let case = () => {
    ignore(Merkle.tree(leaves));
  };
  let name = "tree-generation:sha256 | height = " ++ string_of_int(height);
  Test.create(case, ~name);
};

let compute = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let tree = Merkle.tree(leaves);
  let index = PRNG.random_int(length);
  let leaf = List.nth(leaves, index);
  let case = () => {
    ignore(Merkle.path(~tree, ~leaf));
  };
  let name = "path-computation:sha256 | height = " ++ string_of_int(height);
  Test.create(case, ~name);
};

let verify = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let tree = Merkle.tree(leaves);
  let index = PRNG.random_int(length);
  let leaf = List.nth(leaves, index);
  let path = Merkle.path(~tree, ~leaf);
  let root = Merkle.root(tree);
  let case = () => {
    assert(Merkle.verify(~root, ~leaf, ~path));
  };
  let name = "path-verification:sha256 | height = " ++ string_of_int(height);
  Test.create(case, ~name);
};

let generations = List.map(generate, heights);
let computations = List.map(compute, heights);
let verifications = List.map(verify, heights);

let suite = generations @ computations @ verifications;

Command.run(make_command(suite));
