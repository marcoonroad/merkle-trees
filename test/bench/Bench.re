open Core_bench.Bench;
module Command = Core.Command;
module PRNG = Support.PRNG;
module Merkle = Support.MerkleTree;

let heights = [3, 5, 10, 12, 15];

let power_of_two = number => {
  int_of_float(2.0 ** float_of_int(number));
};

let generate = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let case = () => {
    ignore(Merkle.tree(leaves));
  };
  let name = "merkle-trees tree generation (sha256 / h = " ++ string_of_int(height) ++ ")";
  Test.create(case, ~name);
};

let compute = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let tree = Merkle.tree(leaves);
  let index = PRNG.random_int(length);
  let leaf = List.nth(leaves, index);
  let case = () => {
    ignore(Merkle.path(~tree, ~leaf));
  };
  let name = "merkle-trees path computation (sha256 / h = " ++ string_of_int(height) ++ ")";
  Test.create(case, ~name);
};

let verify = height => {
  let length = power_of_two(height);
  let leaves = PRNG.list_of_random_256_bits(length);
  let tree = Merkle.tree(leaves);
  let index = PRNG.random_int(length);
  let leaf = List.nth(leaves, index);
  let path = Merkle.path(~tree, ~leaf);
  let root = Merkle.root(tree);
  let case = () => {
    assert(Merkle.verify(~root, ~leaf, ~path));
  };
  let name = "merkle-trees path verification (sha256 / h = " ++ string_of_int(height) ++ ")";
  Test.create(case, ~name);
};

let generations = List.map(generate, heights);
let computations = List.map(compute, heights);
let verifications = List.map(verify, heights);

let suite = generations @ computations @ verifications;

Command.run(make_command(suite));
