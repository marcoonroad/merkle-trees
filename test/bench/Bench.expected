Estimated testing time 5m (30 benchmarks x 10s). Change using '-quota'.
merkle-trees tree generation (SHA256 / h = 3): Total time taken 10.01889967918396s (160 samples, max runs 160).
merkle-trees tree generation (SHA256 / h = 5): Total time taken 10.036258697509766s (148 samples, max runs 148).
merkle-trees tree generation (SHA256 / h = 10): Total time taken 10.247571468353271s (55 samples, max runs 55).
merkle-trees tree generation (SHA256 / h = 12): Total time taken 10.93004298210144s (21 samples, max runs 21).
merkle-trees tree generation (SHA256 / h = 15): Total time taken 12.696697235107422s (4 samples, max runs 4).
merkle-trees tree generation (Blake2B / h = 3): Total time taken 10.033361434936523s (146 samples, max runs 146).
merkle-trees tree generation (Blake2B / h = 5): Total time taken 10.060769319534302s (151 samples, max runs 151).
merkle-trees tree generation (Blake2B / h = 10): Total time taken 10.119712591171265s (62 samples, max runs 62).
merkle-trees tree generation (Blake2B / h = 12): Total time taken 10.075368165969849s (22 samples, max runs 22).
merkle-trees tree generation (Blake2B / h = 15): Total time taken 12.603146076202393s (4 samples, max runs 4).
merkle-trees path computation (SHA256 / h = 3): Total time taken 10.000168561935425s (163 samples, max runs 163).
merkle-trees path computation (SHA256 / h = 5): Total time taken 10.018798589706421s (155 samples, max runs 155).
merkle-trees path computation (SHA256 / h = 10): Total time taken 10.17178201675415s (57 samples, max runs 57).
merkle-trees path computation (SHA256 / h = 12): Total time taken 10.443514108657837s (22 samples, max runs 22).
merkle-trees path computation (SHA256 / h = 15): Total time taken 13.032346725463867s (4 samples, max runs 4).
merkle-trees path computation (Blake2B / h = 3): Total time taken 10.057705640792847s (162 samples, max runs 162).
merkle-trees path computation (Blake2B / h = 5): Total time taken 10.050476312637329s (158 samples, max runs 158).
merkle-trees path computation (Blake2B / h = 10): Total time taken 10.198746919631958s (77 samples, max runs 77).
merkle-trees path computation (Blake2B / h = 12): Total time taken 10.165380239486694s (24 samples, max runs 24).
merkle-trees path computation (Blake2B / h = 15): Total time taken 13.250980854034424s (4 samples, max runs 4).
merkle-trees path verification (SHA256 / h = 3): Total time taken 10.045356512069702s (164 samples, max runs 164).
merkle-trees path verification (SHA256 / h = 5): Total time taken 10.01335597038269s (156 samples, max runs 156).
merkle-trees path verification (SHA256 / h = 10): Total time taken 10.033257007598877s (162 samples, max runs 162).
merkle-trees path verification (SHA256 / h = 12): Total time taken 10.045492649078369s (161 samples, max runs 161).
merkle-trees path verification (SHA256 / h = 15): Total time taken 10.027318954467773s (134 samples, max runs 134).
merkle-trees path verification (Blake2B / h = 3): Total time taken 10.027338027954102s (153 samples, max runs 153).
merkle-trees path verification (Blake2B / h = 5): Total time taken 10.006686210632324s (168 samples, max runs 168).
merkle-trees path verification (Blake2B / h = 10): Total time taken 10.0361168384552s (167 samples, max runs 167).
merkle-trees path verification (Blake2B / h = 12): Total time taken 10.025506734848022s (168 samples, max runs 168).
merkle-trees path verification (Blake2B / h = 15): Total time taken 10.012085437774658s (167 samples, max runs 167).
┌───────────────────────────────────────────────────┬────────────────┬────────────────┬────────────┬─────────────┬─────────────┬─────────────┬─────────────┬────────────┐
│ Name                                              │       Time/Run │        mWd/Run │ mWd Overhd │    mjWd/Run │ mjWd Overhd │    Prom/Run │ Prom Overhd │ Percentage │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (SHA256 / h = 3)     │        26.98us │      1_540.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (SHA256 / h = 5)     │       113.49us │      6_120.08w │     24.87w │       6.30w │     -49.03w │       6.30w │     -49.03w │      0.01% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (SHA256 / h = 10)    │     4_731.19us │    194_651.87w │     23.64w │   4_171.03w │  -7_185.44w │   4_171.03w │  -7_185.44w │      0.22% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (SHA256 / h = 12)    │    45_930.73us │    778_355.58w │     30.58w │  43_218.37w │ -62_368.52w │  43_218.37w │ -62_368.52w │      2.11% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (SHA256 / h = 15)    │ 2_073_446.43us │  6_226_078.90w │     19.40w │ 551_181.50w │     857.00w │ 551_181.50w │     857.00w │     95.42% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (Blake2B / h = 3)    │        13.87us │      2_908.00w │     27.00w │       0.75w │     -25.19w │       0.75w │     -25.19w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (Blake2B / h = 5)    │        48.20us │     11_616.00w │     27.00w │      11.99w │    -192.54w │      11.99w │    -192.54w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (Blake2B / h = 10)   │     2_994.31us │    370_770.37w │     28.49w │  11_344.32w │      81.15w │  11_344.32w │      81.15w │      0.14% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (Blake2B / h = 12)   │    36_693.95us │  1_482_855.72w │     22.46w │  79_512.10w │   5_391.86w │  79_512.10w │   5_391.86w │      1.69% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees tree generation (Blake2B / h = 15)   │ 2_043_011.70us │ 11_862_156.70w │     26.20w │ 822_729.90w │   2_611.40w │ 822_729.90w │   2_611.40w │     94.02% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (SHA256 / h = 3)    │         9.95us │        559.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (SHA256 / h = 5)    │        38.00us │      1_887.00w │     27.00w │       0.34w │     -16.41w │       0.34w │     -16.41w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (SHA256 / h = 10)   │     3_877.87us │     55_535.14w │     29.25w │     590.95w │   1_975.87w │     590.95w │   1_975.87w │      0.18% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (SHA256 / h = 12)   │    36_799.54us │    221_456.58w │     24.60w │  10_457.31w │    -370.58w │  10_457.31w │    -370.58w │      1.69% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (SHA256 / h = 15)   │ 2_123_775.62us │  1_769_797.40w │     26.90w │ 266_866.50w │  -2_686.00w │ 266_866.50w │  -2_686.00w │     97.74% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (Blake2B / h = 3)   │         4.81us │      1_013.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (Blake2B / h = 5)   │        16.92us │      3_469.00w │     27.00w │       2.59w │     -63.26w │       2.59w │     -63.26w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (Blake2B / h = 10)  │     1_848.21us │    103_741.02w │     29.84w │   1_839.01w │     -21.57w │   1_839.01w │     -21.57w │      0.09% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (Blake2B / h = 12)  │    31_269.78us │    414_045.14w │     26.07w │  27_568.29w │   1_004.95w │  27_568.29w │   1_004.95w │      1.44% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path computation (Blake2B / h = 15)  │ 2_172_966.38us │  3_309_965.90w │     26.40w │ 399_898.60w │   2_151.60w │ 399_898.60w │   2_151.60w │    100.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (SHA256 / h = 3)   │         6.03us │        327.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (SHA256 / h = 5)   │         8.32us │        429.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (SHA256 / h = 10)  │        13.30us │        684.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (SHA256 / h = 12)  │        15.49us │        786.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (SHA256 / h = 15)  │        22.07us │        939.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (Blake2B / h = 3)  │         3.01us │        632.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (Blake2B / h = 5)  │         3.83us │        828.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (Blake2B / h = 10) │         6.07us │      1_318.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (Blake2B / h = 12) │         6.90us │      1_514.00w │     27.00w │    0.00e-9w │    0.00e-9w │    0.00e-9w │    0.00e-9w │      0.00% │
├───────────────────────────────────────────────────┼────────────────┼────────────────┼────────────┼─────────────┼─────────────┼─────────────┼─────────────┼────────────┤
│ merkle-trees path verification (Blake2B / h = 15) │         8.30us │      1_808.00w │     27.00w │       0.14w │      -7.33w │       0.14w │      -7.33w │      0.00% │
└───────────────────────────────────────────────────┴────────────────┴────────────────┴────────────┴─────────────┴─────────────┴─────────────┴─────────────┴────────────┘
Benchmarks that take 1ns to 100ms can be estimated precisely. For more reliable 
estimates, redesign your benchmark to have a shorter execution time.
